### PR TITLE
rgw: add PublicAccessBlock set of APIs on buckets

### PR DESCRIPTION
**Pull request by theanalyst**
implementation attempt of https://docs.aws.amazon.com/AmazonS3/latest/dev/access-control-block-public-access.html for bucket apis 

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by theanalyst** [comment]
So an initial implementation covering BlockPublicPolicy allowing for Blocking Public ACLs, Policies and Ignoring Public ACLs for *buckets* is ready. RestrictPublicACLs is not specially implemented as I'm not aware how we do cross tenant access in RGW just yet. The tests at https://github.com/ceph/s3-tests/pull/302 should showcase some basic scenarios and how blocking works. With respect to bucket policies we're still doing basic block access but there might be policies with Deny that probably pass through still. Any suggestions on improvements here are welcome. 
 
 At a user level this implementation is still incomplete, as S3 has a s3-config set of service apis for doing this, but our current model maps Accounts to tenants, which don't seem to have a physical entity of sorts where policies can be attached so this needs to be done. A future direction could be that we just map BlockPublicAccess settings to users themselves which should be fairly simple to implement with this framework, just adding another user attribute similar to user policies and using the logic in place. I'll plan to open another PR later once we decide on how to take our stance there.

File: src/rgw/rgw_acl.cc [hunk 1]
```diff
@@ -116,7 +116,8 @@ uint32_t RGWAccessControlList::get_referer_perm(const uint32_t current_perm,
 uint32_t RGWAccessControlPolicy::get_perm(const DoutPrefixProvider* dpp,
                                           const rgw::auth::Identity& auth_identity,
                                           const uint32_t perm_mask,
-                                          const char * const http_referer)
+                                          const char * const http_referer,
+                                          bool IgnorePublicACLs)
```
Comment (cbodley) [inline src/rgw/rgw_acl.cc]: can you please stick to snake_case for variable- and function names? applies to IsPublic() below also

File: src/rgw/rgw_public_access.cc [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  RGWXMLDecoder::decode_xml("RestrictPublicBuckets", RestrictPublicBuckets, obj);
+  //  RGWXMLDecoder::decode_xml("BlockPublicAccessConfiguration", this, obj);
+}
+
+void PublicAccessConfiguration::dump_xml(Formatter *f) const {
+  Formatter::ObjectSection os(*f, "BlockPublicAccessConfiguration");
+  // AWS spec mentions the values to be ALL CAPs, but clients will not
+  // understand this or a mixed case like it is supposed to, hence the need to
+  // manually encode here
+  auto bool_val = [](bool b) -> auto { return b ? "true": "false"; };
```
Comment (cbodley) [inline src/rgw/rgw_public_access.cc]: it looks like Formatter::dump_bool() does exactly this

File: src/rgw/rgw_public_access.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ */
+
+#pragma once
+#include <include/types.h>
+
+class XMLObj;
+
+namespace rgw::IAM {
+
+class PublicAccessConfiguration {
```
Comment (cbodley) [inline src/rgw/rgw_public_access.h]: is PublicAccessConfiguration different than the PublicAccessBlockConfiguration described in s3 docs? https://docs.aws.amazon.com/AmazonS3/latest/API/API_PublicAccessBlockConfiguration.html

if not, can we rename this to PublicAccessBlockConfiguration and move it out of the IAM namespace?

File: src/rgw/rgw_public_access.cc [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#include "rgw_xml.h"
+
+namespace rgw::IAM {
+
+void PublicAccessConfiguration::decode_xml(XMLObj *obj) {
+  RGWXMLDecoder::decode_xml("BlockPublicAcls", BlockPublicAcls, obj);
+  RGWXMLDecoder::decode_xml("IgnorePublicAcls", IgnorePublicAcls, obj);
+  RGWXMLDecoder::decode_xml("BlockPublicPolicy", BlockPublicPolicy, obj);
+  RGWXMLDecoder::decode_xml("RestrictPublicBuckets", RestrictPublicBuckets, obj);
+  //  RGWXMLDecoder::decode_xml("BlockPublicAccessConfiguration", this, obj);
```
Comment (cbodley) [inline src/rgw/rgw_public_access.cc]: remove comment?

File: src/rgw/rgw_rest_s3.cc [hunk 5]
```diff
@@ -3552,6 +3552,47 @@ void RGWGetObjLegalHold_ObjStore_S3::send_response()
...
+  dump_errno(s);
+  end_header(s, this, "application/xml");
+  dump_start(s);
+
+  s->formatter->open_object_section_in_ns("PolicyStatus", XMLNS_AWS_S3);
+  // https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGETPolicyStatus.html
+  // mentions TRUE and FALSE, but boto/aws official clients seem to want lower
+  // case which is returned by AWS as well; so let's be bug to bug compatible
+  // with the API
+  s->formatter->dump_string("IsPublic", isPublic ? "true" : "false");
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc]: dump_bool()

File: src/rgw/rgw_op.h [hunk 6]
```diff
@@ -2345,6 +2346,59 @@ class RGWGetClusterStat : public RGWOp {
...
+class RGWGetBucketPolicyStatus : public RGWOp {
+protected:
+  bool isPublic {false};
+public:
+  int verify_permission() override;
+  const char* name() const override { return "get_bucket_policy_status"; }
+  virtual RGWOpType get_type() override { return RGW_OP_GET_BUCKET_POLICY_STATUS; }
+  virtual uint32_t op_mask() override { return RGW_OP_TYPE_READ; }
+  void execute() override;
+  dmc::client_id dmclock_client() override { return dmc::client_id::metadata; }
```
Comment (cbodley) [inline src/rgw/rgw_op.h]: nit: metadata is the default so it's not necessary to override these

File: src/rgw/rgw_op.cc [hunk 7]
```diff
@@ -303,6 +304,24 @@ static boost::optional<Policy> get_iam_policy_from_attr(CephContext* cct,
   }
 }
 
+static boost::optional<PublicAccessConfiguration>
+get_public_access_conf_from_attr(const map<string, bufferlist>& attrs)
+{
+  if (auto aiter = attrs.find(RGW_ATTR_PUBLIC_ACCESS);
+      aiter != attrs.end())
+    {
```
Comment (cbodley) [inline src/rgw/rgw_op.cc]: there's extra indentation on this block

**Comment by cbodley** [comment]
> With respect to bucket policies we're still doing basic block access but there might be policies with Deny that probably pass through still. Any suggestions on improvements here are welcome.

@adamemerson i'd love your help evaluating this part

File: src/rgw/rgw_public_access.cc (hunk 2 above)
Comment (theanalyst) [reply src/rgw/rgw_public_access.cc]: thanks will change, encode_xml seemed to have bool with a title case that was breaking s3tests even though it shouldn't :/

File: src/rgw/rgw_public_access.h (hunk 3 above)
Comment (theanalyst) [reply src/rgw/rgw_public_access.h]: it is the same thing, I'll change.

File: src/rgw/rgw_iam_policy.cc [hunk 8]
```diff
@@ -1443,14 +1443,13 @@ struct IsPublicStatement
     if (s.effect == Effect::Allow) {
       for (const auto& p : s.princ) {
 	if (p.is_wildcard()) {
-	  if (s.eval_conditions(iam_all_env) == Effect::Allow)
-	    return true;
+	  return s.eval_conditions(iam_all_env) == Effect::Allow;
 	}
       }
       // no princ should not contain fixed values
-      return std::all_of(s.noprinc.begin(), s.noprinc.end(), [](const rgw::auth::Principal& p) {
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_iam_policy.cc]: @adamemerson interesting

[the remark this answers is not in the document]
Comment (theanalyst) [reply src/rgw/rgw_iam_policy.cc]: Good point, they can coexist,but wouldn't an allow with a specific value and deny in the same policy make it non-public, unless the allow was a wildcard?

[the remark this answers is not in the document]
Comment (adamemerson) [reply src/rgw/rgw_iam_policy.cc]: Yes. It would be non-public in that case.

**Comment by theanalyst** [comment]
jenkins test this please

**Comment by theanalyst** [comment]
Are there any comments that needs addressing for the bucket level apis atm?

**Comment by cbodley** [comment]
can we make sure to run qa against https://github.com/ceph/s3-tests/pull/302?

**Comment by theanalyst** [comment]
> can we make sure to run qa against [ceph/s3-tests#302](https://github.com/ceph/s3-tests/pull/302)?

sure will do that

**Comment by theanalyst** [comment]
jenkins test this please

**Comment by yuriw** [comment]
[wip-yuri8-testing-2019-12-11-2105](https://trello.com/c/QPjkDkCb/778-wip-yuri8-testing-2019-12-11-2105)

**Comment by theanalyst** [comment]
tests at http://pulpito.ceph.com/abhi-2020-01-15_10:46:20-rgw-wip-abhi-testing-2020-01-13-1517-distro-basic-smithi/ 
s3 test failures were because of a ceph-master branch, so these passed at http://pulpito.ceph.com/abhi-2020-01-22_16:08:45-rgw-wip-abhi-testing-2020-01-13-1517-distro-basic-smithi/ 

seeing failures at hadoop jobs, and rgw-admin tasks with subuser rm (shouldn't be related), Is this good to go in?

**Comment by theanalyst** [comment]
After rebase on master, this passes s3-tests 
run: http://pulpito.ceph.com/abhi-2020-01-24_16:05:12-rgw-wip-abhi-testing-2020-01-24-1135-distro-basic-smithi/ (which had s3-tests failure because of not being based on ceph-master) 
rerun: http://pulpito.ceph.com/abhi-2020-01-27_10:30:43-rgw-wip-abhi-testing-2020-01-24-1135-distro-basic-smithi/
the current failures seen there are s3-website (centos), multisite failures and 3 other failures that seems to be environmental (paramiko failed to connect?)

**Comment by theanalyst** [comment]
There is a change introduced in the last commit after the bucket granularity sync which introduced `perm_state_base` which now also has a member for public access config. @cbodley @yehudasa

File: src/rgw/rgw_common.h [hunk 9]
```diff
@@ -2121,7 +2132,24 @@ struct perm_state_base {
...
+                                                defer_to_bucket_acls(_defer_to_bucket_acls)
+  {}
+
+  perm_state_base(CephContext *_cct,
+                  const rgw::IAM::Environment& _env,
+                  rgw::auth::Identity *_identity,
+                  const RGWBucketInfo& _bucket_info,
+                  int _perm_mask,
+                  bool _defer_to_bucket_acls,
+                  boost::optional<PublicAccessBlockConfiguration> _bucket_acess_conf) : cct(_cct),
```
Comment (cbodley) [inline src/rgw/rgw_common.h]: this argument should just be added to the existing constructor with a default value, ie `_bucket_access_conf = boost::none`

File: src/rgw/rgw_common.cc [hunk 10]
```diff
@@ -848,7 +848,9 @@ void RGWHTTPArgs::append(const string& name, const string& val)
       (name.compare("torrent") == 0) ||
       (name.compare("tagging") == 0) ||
       (name.compare("append") == 0) ||
-      (name.compare("position") == 0)) {
+      (name.compare("position") == 0) ||
+      (name.compare("policyStatus") == 0) ||
+      (name.compare("publicAccessBlock"))) {
```
Comment (cbodley) [inline src/rgw/rgw_common.cc]: name.compare("publicAccessBlock") == 0

**Comment by theanalyst** [comment]
test run with updated branch came out green (except 2 multiste/pubsub failures) http://pulpito.ceph.com/abhi-2020-02-03_19:59:36-rgw-wip-abhi-testing-2020-02-03-1754-distro-basic-smithi/

**Comment by theanalyst** [comment]
@cbodley @adamemerson any final comments, I'll create a docs pr, is this good to merge otherwise

**Comment by skuffe** [comment]
I've been attempting to use the openshift image registry operator with s3 compatible storage backed by ceph rgw, and i'm running into some issues with the PublicAccessBlock. If i let the operator create a bucket on it's own it will apply a PublicAccessBlock with the following config;

BlockPublicAcls: true
BlockPublicPolicy: true
IgnorePublicAcls: true
RestrictPublicBuckets: true

This seems to effectively block any attempts to put new objects into the bucket (403 access denied response), even when authenticated as the bucket owner.

As soon as i remove the public access block, e.g `s3api delete-public-access-block --bucket registry` i can put objects into the bucket again.

Am i missing something or could this be a bug?

Manual tests were done using the aws cli tool.
Ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)

**Pull request merged**